lib/Sema/CSSolver.cpp

```diff
@@ -1503,21 +1502,66 @@ static bool havePotentialTypesOrLiteralConformances(Type ty,
       case ConstraintKind::FunctionResult:
       case ConstraintKind::SelfObjectOfProtocol:
       case ConstraintKind::ConformsTo:
+      case ConstraintKind::Defaultable:
         break;
       }
     }
   }
+}
 
-  return false;
+void ConstraintSystem::ArgumentInfoCollector::minimizeLiteralProtocols() {
+  if (LiteralProtocols.size() <= 1)
+    return;
+
+  llvm::SmallVector<Type, 2> defaultTypes;
+  for (auto *protocol : LiteralProtocols)
+    defaultTypes.push_back(CS.TC.getDefaultType(protocol, CS.DC));
+
+  auto result = 0;
+  for (unsigned long i = 1; i < LiteralProtocols.size(); ++i) {
+    auto first =
+        CS.TC.conformsToProtocol(defaultTypes[i], LiteralProtocols[result],
+                                 CS.DC, ConformanceCheckFlags::InExpression);
+    auto second =
+        CS.TC.conformsToProtocol(defaultTypes[result], LiteralProtocols[i],
+                                 CS.DC, ConformanceCheckFlags::InExpression);
+    if ((first && second) || (!first && !second))
+      return;
+
+    if (first)
+      result = i;
+  }
+
+  auto *protocol = LiteralProtocols[result];
+  LiteralProtocols.clear();
+  LiteralProtocols.insert(protocol);
+}
+
+void ConstraintSystem::ArgumentInfoCollector::dump() const {
+  auto &log = CS.getASTContext().TypeCheckerDebug->getStream();
+  log << "types:\n";
+  for (auto type : Types)
+    type->print(log);
+  log << "\n";
+
+  log << "literal protocols:\n";
+  for (auto *proto : LiteralProtocols)
+    proto->print(log);
+  log << "\n";
 }
 
 // Check to see if we know something about the types of all arguments
 // in the given function type.
-static bool haveTypeInformationForAllArguments(AnyFunctionType *fnType,
-                                               ConstraintSystem &cs) {
-  return llvm::all_of(fnType->getParams(), [&](AnyFunctionType::Param param) {
-    return havePotentialTypesOrLiteralConformances(param.getPlainType(), cs);
-  });
+bool ConstraintSystem::haveTypeInformationForAllArguments(
+    AnyFunctionType *fnType) {
+  llvm::SetVector<Constraint *> literalConformsTo;
+  return llvm::all_of(fnType->getParams(),
+                      [&](AnyFunctionType::Param param) -> bool {
+                        ArgumentInfoCollector argInfo(*this, param);
+                        return argInfo.getTypes().size() +
```

Comment:
The formatting here is a bit hard to read

Reply:
Yeah that's unfortunate and I think just the result of `clang-format`ing it. I can take a look at improving it.